diff -ruN '--exclude=.git' '--exclude=obj' avr-libc/include/iconv.h avr-libc-new/include/iconv.h
--- avr-libc/include/iconv.h	1969-12-31 16:00:00.000000000 -0800
+++ avr-libc-new/include/iconv.h	2024-05-24 18:29:32.490313526 -0700
@@ -0,0 +1,12 @@
+#ifndef __ICONV_H_
+#define __ICONV_H_ 1
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef __cplusplus
+}
+#endif
+#endif
diff -ruN '--exclude=.git' '--exclude=obj' avr-libc/include/Makefile.am avr-libc-new/include/Makefile.am
--- avr-libc/include/Makefile.am	2024-05-25 15:59:55.018271496 -0700
+++ avr-libc-new/include/Makefile.am	2024-05-24 15:46:07.750998711 -0700
@@ -47,7 +47,8 @@
     stdlib.h \
     string.h \
     time.h \
-    unistd.h
+    unistd.h \
+    iconv.h
 
 
 SUBDIRS = avr compat util sys
